barretenberg/cpp/src/barretenberg/vm2/constraining/relations/emit_nullifier.test.cpp

```diff
@@ -0,0 +1,243 @@
+    TestTraceContainer trace({ {
+        { C::execution_sel_execute_emit_nullifier, 1 },
+        { C::execution_register_0_, /*nullifier=*/42 },
+        { C::execution_mem_tag_reg_0_, static_cast<uint8_t>(MemoryTag::FF) },
+        { C::execution_remaining_nullifiers_inv, 0 },
+        { C::execution_sel_write_nullifier, 0 },
+        { C::execution_sel_opcode_error, 1 },
+        { C::execution_subtrace_operation_id, AVM_EXEC_OP_ID_EMIT_NULLIFIER },
+        { C::execution_prev_nullifier_tree_size, 1 },
+        { C::execution_nullifier_tree_size, 1 },
+        { C::execution_prev_num_nullifiers_emitted, prev_num_nullifiers_emitted },
+        { C::execution_num_nullifiers_emitted, prev_num_nullifiers_emitted },
+    } });
+    check_relation<emit_nullifier>(trace);
+
+    // Negative test: sel_opcode_error must be on
+    trace.set(C::execution_sel_opcode_error, 0, 0);
+    EXPECT_THROW_WITH_MESSAGE(check_relation<emit_nullifier>(trace, emit_nullifier::SR_OPCODE_ERROR_IF_LIMIT_ERROR),
+                              "OPCODE_ERROR_IF_LIMIT_ERROR");
+
+    // Negative test: tree size must be the same
+    trace.set(C::execution_nullifier_tree_size, 0, 2);
+    EXPECT_THROW_WITH_MESSAGE(
+        check_relation<emit_nullifier>(trace, emit_nullifier::SR_EMIT_NULLIFIER_TREE_SIZE_INCREASE),
+        "EMIT_NULLIFIER_TREE_SIZE_INCREASE");
+
+    // Negative test: num nullifiers emitted must be the same
+    trace.set(C::execution_num_nullifiers_emitted, 0, prev_num_nullifiers_emitted + 1);
+    EXPECT_THROW_WITH_MESSAGE(
+        check_relation<emit_nullifier>(trace, emit_nullifier::SR_EMIT_NULLIFIER_NUM_NULLIFIERS_EMITTED_INCREASE),
+        "EMIT_NULLIFIER_NUM_NULLIFIERS_EMITTED_INCREASE");
+}
+
+TEST(EmitNullifierConstrainingTest, Interactions)
+{
+    NiceMock<MockPoseidon2> poseidon2;
+    NiceMock<MockMerkleCheck> merkle_check;
+    NiceMock<MockRangeCheck> range_check;
+    NiceMock<MockFieldGreaterThan> field_gt;
+
+    EventEmitter<NullifierTreeCheckEvent> nullifier_tree_check_event_emitter;
+    NullifierTreeCheck nullifier_tree_check(poseidon2, merkle_check, field_gt, nullifier_tree_check_event_emitter);
+
+    FF nullifier = 42;
+    AztecAddress contract_address = 0xdeadbeef;
+    FF siloed_nullifier = 66;
+    FF low_leaf_nullifier = 99; // siloed nullifier != low leaf nullifier (NO COLLISION)
+    FF low_leaf_hash = 77;
+    FF updated_low_leaf_hash = 101;
+    FF new_leaf_hash = 111;
+    FF pre_write_root = 27;
+    FF intermediate_root = 33;
+    FF post_write_root = 88;
+
+    // insertion sibling path
+    std::vector<FF> insertion_sibling_path = { 1, 2, 3 };
+
+    AppendOnlyTreeSnapshot prev_snapshot = AppendOnlyTreeSnapshot{
+        .root = pre_write_root,
+        .nextAvailableLeafIndex = 128,
+    };
+    uint32_t prev_num_nullifiers_emitted = 2;
+
+    // mock the nullifier > low leaf nullifier to return true
+    EXPECT_CALL(field_gt, ff_gt).WillOnce(Return(true));
```

Comment:
not necessarily for this PR but looks like a FakeFieldGT might be useful at some point later